NQuads.php: made blank node regex more flexible

k00ni opened the pull request:
As far as I understand https://www.w3.org/TR/n-quads/#BNodes, blank node labels in NQuads can include `-`, `.` and `_` too. Specification requires these characters not being at the beginning of the label, but this regex is not structured to check that anyway.

k00ni committed:
NQuads.php: made blank node regex more flexible

As far as I understand https://www.w3.org/TR/n-quads/#BNodes, blank node labels in NQuads can include -, . and _ too. Specification requires these characters not being at the beginning of the label, but this regex is not structured to check that anyway.

lanthaler commented:
The current regex is too restrictive as you point out. The proposed new one is too liberal. Would you mind fixing it to match the spec?

k00ni committed:
made bnode regex more strict again, but still include: - . _

k00ni commented:
I updated the regex and added a few tests to prove they work. Is that approach OK?

Any idea how to handle the following characters? U+00B7, U+0300 to U+036F and U+203F to U+2040

lanthaler commented:
You should be able to match then with `\x{00B7}` etc. If I read the regex in the change correctly, it would disallow IDs such as `_:b` as it always requires at least two characters. You'll have to make the second oder optional.

k00ni committed:
include edge case (_:b) // include missing allowed UTF-8 characters

k00ni commented:
`_:b` is covered now.

To include  U+00B7, U+0300 to U+036F and U+203F to U+2040, I had to use UTF-8 mode in regex. But after doing it, a few tests fail, because at [Processor.php line 2895](https://github.com/lanthaler/JsonLD/blob/master/Processor.php#L2895) it tries to use an non-existing property, like:

```
1) ML\JsonLD\Test\NQuadsTest::testEscaping
Error: Cannot access empty property

/var/www/html/Processor.php:2895
/var/www/html/Processor.php:2322
/var/www/html/JsonLD.php:465
/var/www/html/Test/NQuadsTest.php:45
phpvfscomposer:///var/www/html/vendor/phpunit/phpunit/phpunit:51
```

That part around line 2895 is weird. First the code:

```php
if (property_exists($object, $property)) {
    if (false === is_array($object->{$property})) {
        $object->{$property} = array($object->{$property});
    }

    if ($unique) {
        foreach ($object->{$property} as $item) {
            if (self::subtreeEquals($item, $value)) {
                return;
            }
        }
    }

    $object->{$property}[] = $value;
} else {
    $object->{$property} = ($alwaysArray) ? array($value) : $value;  // <--------------- faulty line 2895
}
```

In the if clause it checks if `$object` has a given `$property`. If it doesn't it goes in the else-part, but `$object->{$property}` anyway. What is the meaning of this?

We have two options as far as I can tell:
1. Either fix this "behavior" which leads to this error in `Processor.php` and keep UTF-8 characters.
2. Or don't include UTF-8 characters in regex.

WDYT?

k00ni commented:
How do we proceed here @lanthaler?

lanthaler commented:
Thanks for the ping, I missed your previous comment. If suggest we remove the UTF-8 characters for now. We can split that of into a separate bug.

k00ni committed:
removed UTF-8 extension; refined tests

k00ni committed:
removed obsolete u in regex

k00ni commented:
I wanted to revert my UTF-8 changes but after running PHPUnit (both **PHP 5.6 and 7.4**), I ran into a couple of errors and failures, such as:

```
1) ML\JsonLD\Test\W3CTestSuiteTest::testRemoteDocumentLoading with data set "http://json-ld.org/test-suite/tests/remote-doc-manifest.jsonld#t0012" ('Multiple context link headers', stdClass Object (...), stdClass Object (...))
Failed asserting that 'loading document failed' is equal to expected exception code 'multiple context link headers'.

phpvfscomposer:///var/www/html/vendor/phpunit/phpunit/phpunit:51
```
or
```
10) ML\JsonLD\Test\W3CTestSuiteTest::testRemoteDocumentLoading with data set "http://json-ld.org/test-suite/tests/remote-doc-manifest.jsonld#t0011" ('load JSON document with exten...h link', stdClass Object (...), stdClass Object (...))
ML\JsonLD\Exception\JsonLdException: Unable to load the remote document "https://json-ld.org:443/test-suite/tests/remote-doc-0011-in.jldt".

/var/www/html/FileGetContentsLoader.php:62
/var/www/html/JsonLD.php:142
/var/www/html/Test/W3CTestSuiteTest.php:182
phpvfscomposer:///var/www/html/vendor/phpunit/phpunit/phpunit:51
```
My changes **are not** the source of these errors because they persisted after reverting ALL of my changes. Unfortunately, there is no CI for pull requests to have a neutral view on the matter. I assume there were changes http://json-ld.org/test-suite/tests/remote-doc-manifest.jsonld in the meantime which broke your tests.

I leave it to you how to proceed.

lanthaler commented:
Oh, I didn't realize that CI is broken. Travis was running on PR before. Your code LGTM. I'll have a look at the tests and the CI setup (likely on the weekend) and merge your code.

Thanks a lot for your work on this Konrad.

k00ni commented:
> [...] and the CI setup (likely on the weekend)

I recommend to use Github Actions instead of Travis.

In the past I had similar problems like these guys ([ref](https://www.nymedia.no/en/blog/why-we-decided-migrate-from-travis-to-github-action), [ref2](https://www.codetd.com/en/article/12116056)). Basically, their service was not reliable anymore.

lanthaler commented:
Could you please [allow edits to this PR](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)? I've found the bug and would like to push it back to the PR before merging it to have the full history here

k00ni commented:
> Could you please [allow edits to this PR](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)?

It is allowed already. "Allow edits by maintainers" is checked.

k00ni commented:
Btw. you mean the bug which stopped us from using UTF-8 symbols in the regex?

lanthaler committed:
Fix bnode regex and fix/simplify tests

lanthaler committed:
Remove unnecessary import

lanthaler commented:
No, the bug that caused tests to fail. I didn't look into the UTF-8 symbols yet.

Thanks for working on this!
